### ISSUE TITLE
MXNet getting lower AUC compared to other frameworks in multigpu multilabel problem

### ISSUE THREAD

**Issue by miguelgfierro** [ilkarman/DeepLearningFrameworks#69]
I'm having problems to match MXNet AUC with the other frameworks like Keras, TF or Pytorch. 

In this [notebook](https://github.com/ilkarman/DeepLearningFrameworks/blob/f4eb2c6b1d0a3a4cbd08d32d5464bc32c8c8e50f/notebooks/MXNet_MultiGPU.ipynb) I get an AUC of 0.73, whereas in [Pytorch](https://github.com/ilkarman/DeepLearningFrameworks/blob/f4eb2c6b1d0a3a4cbd08d32d5464bc32c8c8e50f/notebooks/PyTorch_MultiGPU.ipynb) for example I get almost 0.80

Any guidance @ThomasDelteil?

**Comment by ThomasDelteil** [comment]
I'll have a longer look tonight but one thing based on the print statement i see is that it looks like you are using different size of dataset?

(78484, 15)
(11212, 15)
(22424, 15)
vs
Loaded 87306 labels and 87306 images
Loaded 7616 labels and 7616 images
Loaded 17198 labels and 17198 images

It isn't very clear exactly how you pre-processed the data, but it looks like on pytorch you do width+40 then random crop? One thing I would investigate is pre-process some specific images and make sure they are equal, adding one transform at a time.

You also scale the learning rate on the MXNet example and not on the pytorch one.

Also note that on the MXNet notebook your custom cross-entropy loss is computed on CPU because of the .assnumpy() statements. Try to remove them and use the nd.array api instead:

For example:
```
if actually_calculate_loss:
            p = in_data[0].asnumpy()  # shape=(b,d)
            y = in_data[1].asnumpy()
            out = y * np.log(p+self.eps) + (1.-y) * np.log((self.eps1) - p)
            self.assign(out_data[0], req[0], mx.nd.array(out))
```

should be:
```
if actually_calculate_loss:
            p = in_data[0]  # shape=(b,d)
            y = in_data[1]
            out = y * nd.log(p+self.eps) + (1.-y) * nd.log((self.eps1) - p)
            self.assign(out_data[0], req[0], out)
```

**Comment by ilkarman** [comment]
So for [pytorch](https://github.com/ilkarman/DeepLearningFrameworks/blob/master/notebooks/PyTorch_MultiGPU.ipynb)

```
CLASSES = 14
from common.utils import get_imgloc_labels, get_train_valid_test_split
self.img_locs, self.labels = get_imgloc_labels(img_dir, lbl_file, patient_ids)
train_set, valid_set, test_set = get_train_valid_test_split(TOT_PATIENT_NUMBER)

#Loaded 87306 labels and 87306 images
#Loaded 7616 labels and 7616 images
#Loaded 17198 labels and 17198 images
```

Thomas is right

**Comment by ilkarman** [comment]
@ThomasDelteil It's true that there were some errors in the processing so that it wasn't consistent with the others, however unfortunately the error remains. I think it's mainly to do with the custom loss function.

It seems strange mxnet doesn't have a binary CE loss? There is a gluon one; is it possible to slot it into mxnet do you think? https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/loss.py#L176

```
    def __init__(self, from_sigmoid=False, weight=None, batch_axis=0, **kwargs):
        super(SigmoidBinaryCrossEntropyLoss, self).__init__(weight, batch_axis, **kwargs)
        self._from_sigmoid = from_sigmoid

    def hybrid_forward(self, F, pred, label, sample_weight=None):
        label = _reshape_like(F, label, pred)
        if not self._from_sigmoid:
            # We use the stable formula: max(x, 0) - x * z + log(1 + exp(-abs(x)))
            loss = F.relu(pred) - pred * label + F.Activation(-F.abs(pred), act_type='softrelu')
        else:
            loss = -(F.log(pred+1e-12)*label + F.log(1.-pred+1e-12)*(1.-label))
        loss = _apply_weighting(F, loss, self._weight, sample_weight)
        return F.mean(loss, axis=self._batch_axis, exclude=True)
```

**Comment by ThomasDelteil** [comment]
The backward propagation in gluon is done through autograd. There is no explicit backward pass here.
You can write your own custom loss using symbols like that https://mxnet.incubator.apache.org/tutorials/r/CustomLossFunction.html
The doc is actually mentioning cross entropy but not implementing it properly
https://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.MakeLoss

**Comment by ThomasDelteil** [comment]
Ok guys, couldn't sleep and I was trying to figure out a way to make it work, ended up coding it in gluon and got 0.81 AUC 😄 it takes about 7min 48 to train on 4 V100 but there is some margin in terms of utilization on the GPUs, oscillating between 50-70% GPU util. I'll issue a PR with my notebook once I've cleaned it. Also there is a bug in the Gluon DataLoader for the RecordIO that I had to hot fix to make it work. Going to sleep better now! https://github.com/apache/incubator-mxnet/issues/9974

**Comment by ilkarman** [comment]
@ThomasDelteil that's great! I have updated the Keras, TF, Pytorch multi-gpu examples in master if you want to check the newest formatting. I'm going to push Chainer soon but it's not working too well. I'm curious if the Gluon example also allows kvstore to be NCCL?

**Comment by ilkarman** [comment]
BTW, I will issue a PR to master soon where the images are resized to 264, 264 and mode=RGB straight after downloading (the alternative would be to create a common optimised openCV resize method):

```
def resize_im(im_loc, size=(264, 264)):
    # Size 264 allows centre-crop of 224 for image-augmentation
    im = Image.open(loc).resize(size, Image.BILINEAR).convert('RGB')
    im.save(im_loc)


def resize_chestxray_mp(im_list):
        ......
        print("Data Download Complete")
        print("About to overwrite images: Image.open(loc).resize((264, 264), Image.BILINEAR).convert('RGB')")
        resize_chestxray_mp(img_locs)
        print("Finished resizing")  
```

**Comment by ThomasDelteil** [comment]
Sure i'll update to the latest format. I want to try a few things more as well, like weighting more examples that have more labels. I tried with `kvstore='nccl'`, I don't get any perf improvement though, not sure why

**Comment by ilkarman** [comment]
BTW, @ThomasDelteil I'm really curious if MXNet Module API has any example for a multi-class (binary cross entropy) loss. I thought it would be quite popular.

**Comment by ThomasDelteil** [comment]
Yes it seems strange that it's not built-in. I mean the custom loss example does mention it as an example (https://mxnet.incubator.apache.org/tutorials/r/CustomLossFunction.html):
```
cross_entropy = label * mx.sym.log(out) + (1 - label) * mx.sym.log(1 - out)
loss = MakeLoss(cross_entropy)
```
Here's an example of a custom loss:
https://stackoverflow.com/questions/49430560/how-to-use-customized-loss-function-with-mxnet
(As I explained in the question, remember to use a custom metric to simply forward your loss if you are using the high-api)

**Comment by ThomasDelteil** [comment]
Here is the gluon multi gpu notebook: https://github.com/ilkarman/DeepLearningFrameworks/pull/81

**Comment by ilkarman** [comment]
@miguelgfierro When you get back from leave (and I am on leave :) ) we thought of:
i) Adding class-labels to records through flag im2rec script (so that your labels are BATCH,14 not BATCH); otherwise it won't do multi-label properly
ii) Reducing your loss to sum / avergae
